source/pele/base_potential.h

```diff
@@ -15,7 +15,32 @@ namespace pele {
 class BasePotential {
 public:
     virtual ~BasePotential() {}
-
+    
+    /**
+     * Return the number of dimensions (box dimensions).
+     * Ideally this should be overloaded.
+     */
+    virtual size_t get_ndim()
+    {
+        throw std::runtime_error("BasePotential::get_ndim must be overloaded");
+    }
+    
+    /**
+     * Return the distance as measured by the distance policy. 
+     */
+    virtual void get_rij(double * const r_ij, double const * const r1, double const * const r2) const
+    {
+        throw std::runtime_error("BasePotential::get_rij must be overloaded");
+    }
+    
+    /**
+     * Return energy_gradient of iteraction. 
+     */
+    virtual double get_interaction_energy_gradient(double r2, double *gij, size_t atom_i, size_t atom_j) const
```

Comment:
Passing the distance function sounds good. Probably then we might as well also pass the interaction-energy-gradient function? When trying to use SimplePairwisePotential, there were problems because it is templated on the distance policy and the interaction type, so it was complicated to do e.g. dynamic cast, http://www.cplusplus.com/reference/memory/dynamic_pointer_cast/.
Would it make sense to add a default template argument for the interaction (not just the distance policy) in SimplePairwisePotential? Otherwise one can also have the intermediate level Stefano is talking about?